### PR TITLE
write_file: add rule and tests

### PR DESCRIPTION
**Pull request by laszlocsomor**
This PR adds two new rules: write_file and
write_xfile.

Both rules solve a common problem: to write a text
file with user-defined contents.

The problem is routinely solved using a genrule.
That however requires Bash, since genrules execute
Bash commands.  Requiring Bash is a problem on
Windows.

The new rules do not require any shell.

The only difference between the rules is that
write_xfile creates an executable file while
write_file doesn't.

See https://github.com/bazelbuild/bazel/issues/4319

### CODE REVIEW

**Comment by brandjon** [comment]
Two rules, one for non-executable and one for executable? Why not one rule and a boolean attribute? This would be more similar to `ctx.actions.write(..., is_executable=...)`.

**Comment by laszlocsomor** [comment]
I believe that's not possible because the `rule()` function must declare whether it's executable.

**Comment by laszlocsomor** [comment]
I could create a macro that instantiates the right rule type though. WDYT?

**Comment by brandjon** [comment]
I have no strong opinion about a macro vs two rules.

**Comment by thomasvl** [comment]
We did the macro to wedge in the impl detail in apple_support::apple_genrule so when the support exists to clean this up, we could.  Up you you all if you want that here or not.

**Comment by laszlocsomor** [comment]
I'll go with the macro then. Thank you both for the comments!

**Comment by laszlocsomor** [comment]
PTAL -- I merged the two rules into one macro.

**Comment by laszlocsomor** [comment]
```
🛠️ Printing raw output for debugging | 0s
-- | --
  | ##### :bazel: buildifier: error while downloading Buildifier:
  | <pre><code>Unknown Buildifier version &#x27;True&#x27;</code></pre>
```

Looks like an error in Buildifier or the Buildkite pipeline?

**Comment by laszlocsomor** [comment]
@thomasvl , thanks for your tips on adding Stardoc (https://github.com/bazelbuild/bazel-skylib/pull/123#issuecomment-470573930).

Below is the generated documentation.

---

## write_file

<pre>
write_file(<a href="#write_file-name">name</a>, <a href="#write_file-out">out</a>, <a href="#write_file-content">content</a>, <a href="#write_file-is_executable">is_executable</a>, <a href="#write_file-kwargs">kwargs</a>)
</pre>

Creates a UTF-8 encoded text file.

### Parameters

<table class="params-table">
  <colgroup>
    <col class="col-param" />
    <col class="col-description" />
  </colgroup>
  <tbody>
    <tr id="write_file-name">
      <td><code>name</code></td>
      <td>
        required.
        <p>
          Name of the rule.
        </p>
      </td>
    </tr>
    <tr id="write_file-out">
      <td><code>out</code></td>
      <td>
        required.
        <p>
          Path of the output file, relative to this package.
        </p>
      </td>
    </tr>
    <tr id="write_file-content">
      <td><code>content</code></td>
      <td>
        optional. default is <code>[]</code>
        <p>
          Lines of text, the contents of the file. Newlines are added
  automatically after every line except the last one.
        </p>
      </td>
    </tr>
    <tr id="write_file-is_executable">
      <td><code>is_executable</code></td>
      <td>
        optional. default is <code>False</code>
      </td>
    </tr>
    <tr id="write_file-kwargs">
      <td><code>kwargs</code></td>
      <td>
        optional.
      </td>
    </tr>
  </tbody>
</table>

File: tests/write_file/BUILD [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+package(default_testonly = 1)
+
+sh_test(
+    name = "write_file_tests",
+    srcs = ["write_file_tests.sh"],
+    data = [
+        ":run_executables",
+        # Use DefaultInfo.files from 'write_empty_text' (via 'file_deps').
+        ":file_deps",
```
Comment (c-parsons) [inline tests/write_file/BUILD]: There are not alphabetical, which is a sign that you haven't yet run buildifier. :) Be sure to do so (and retain the comment ordering)

File: tests/write_file/BUILD [hunk 2]
```diff
@@ -0,0 +1,83 @@
+load("//rules:write_file.bzl", "write_file")
```
Comment (c-parsons) [inline tests/write_file/BUILD]: Apologies, but I'm really at a bit of a loss for what this BUILD file is accomplishing with its various genrule -> write targets. Perhaps a top-level comment to describe the tree might help with readability?

File: docs/BUILD [hunk 3]
```diff
@@ -4,98 +4,104 @@ load("@io_bazel_skydoc//stardoc:stardoc.bzl", "stardoc")
 # the --incompatible_remap_main_repo flag.
 stardoc(
     name = "build_test_docs",
-    input = "//rules:build_test.bzl",
     out = "build_test_doc_gen.md",
+    input = "//rules:build_test.bzl",
```
Comment (c-parsons) [inline docs/BUILD]: please be sure to pull master and merge :)

File: tests/write_file/BUILD (hunk 1 above)
Comment (laszlocsomor) [reply tests/write_file/BUILD]: I have. Buildifier doesn't reorder these lines.

File: docs/BUILD (hunk 3 above)
Comment (laszlocsomor) [reply docs/BUILD]: Done.

File: tests/write_file/BUILD (hunk 2 above)
Comment (laszlocsomor) [reply tests/write_file/BUILD]: Done.

**Comment by laszlocsomor** [comment]
@laurentlb @brandjon @vladmos : could one of you merge this please?

Would it make sense to grant me merge rights? I'm planning to do more changes in this repo.

**Comment by vladmos** [comment]
I don't seem to have merge rights here as well.

**Comment by laszlocsomor** [comment]
ping -- please help merging

**Comment by c-parsons** [comment]
@vladmos and @laszlocsomor you now both have write / merge privileges.

**Pull request merged**